driver/controller.go

```diff
@@ -963,3 +977,46 @@ func validateCapabilities(caps []*csi.VolumeCapability) bool {
 
 	return supported
 }
+
+func (d *Driver) tagVolume(parentCtx context.Context, vol *godo.Volume) error {
+	for _, tag := range vol.Tags {
+		if tag == d.doTag {
+			return nil
+		}
+	}
+
+	tagReq := &godo.TagResourcesRequest{
+		Resources: []godo.Resource{
+			godo.Resource{
+				ID:   vol.ID,
+				Type: godo.VolumeResourceType,
+			},
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(parentCtx, doAPITimeout)
+	defer cancel()
+	resp, err := d.tags.TagResources(ctx, d.doTag, tagReq)
+
+	if resp == nil || resp.StatusCode != http.StatusNotFound {
+		// either success or irrecoverable failure
+		return err
+	}
+
+	// godo.TagsService returns 404 if the tag has not yet been
+	// created, if that happens we need to create the tag
+	// and then retry tagging the volume resource.
+	ctx, cancel = context.WithTimeout(parentCtx, doAPITimeout)
+	defer cancel()
+	_, _, err = d.tags.Create(parentCtx, &godo.TagCreateRequest{
+		Name: d.doTag,
+	})
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel = context.WithTimeout(parentCtx, doAPITimeout)
+	defer cancel()
+	_, err = d.tags.TagResources(ctx, d.doTag, tagReq)
+	return err
+}
```

Comment:
Seems like instead of setting individual timeouts for each request, we can set the `client` timeout for our default `http.Client`. This way you don't have to set an explicit timeout for each request. To do that, we need to pass a custom http.Client to the oauth2 library:

```go
// the oauth2 pkg only allows us to modify the http.Client via a ctx key
httpClient := &http.Client{Timeout: doAPITimeout}
ctx = context.WithValue(ctx, oauth2.HTTPClient, httpClient)

resp, err := d.tags.TagResources(ctx, d.doTag, tagReq)
if resp == nil || resp.StatusCode != http.StatusNotFound {
	// either success or irrecoverable failure
	return err
}

// godo.TagsService returns 404 if the tag has not yet been
// created, if that happens we need to create the tag
// and then retry tagging the volume resource.
_, _, err = d.tags.Create(parentCtx, &godo.TagCreateRequest{
	Name: d.doTag,
})
if err != nil {
	return err
}

_, err = d.tags.TagResources(ctx, d.doTag, tagReq)
return err
```

Reply:
@fatih and I discussed this.  The above approach may not viable upon closer inspection.  https://github.com/golang/oauth2/blob/master/oauth2.go#L348 seems to show that the http.Client is only extracted from the context in `oauth2.NewClient`.